// nivo/es/components/charts/sankey/SankeyNodesItem.js
/*
 * This file is part of the nivo project.
 *
 * Copyright 2016-present, Raphaël Benitte.
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */
import React from 'react';
import PropTypes from 'prop-types';
import compose from 'recompose/compose';
import withState from 'recompose/withState';
import withPropsOnChange from 'recompose/withPropsOnChange';
import withHandlers from 'recompose/withHandlers';
import pure from 'recompose/pure';
import BasicTooltip from '../../tooltip/BasicTooltip';

var SankeyNodesItem = function SankeyNodesItem(_ref) {
    var node = _ref.node,
        x = _ref.x,
        y = _ref.y,
        width = _ref.width,
        height = _ref.height,
        color = _ref.color,
        opacity = _ref.opacity,
        borderWidth = _ref.borderWidth,
        borderColor = _ref.borderColor,
        handleMouseEnter = _ref.handleMouseEnter,
        handleMouseMove = _ref.handleMouseMove,
        handleMouseLeave = _ref.handleMouseLeave,
        onClick = _ref.onClick;
    return React.createElement('rect', {
        x: x,
        y: y,
        width: width,
        height: height,
        fill: color,
        fillOpacity: opacity,
        strokeWidth: borderWidth,
        stroke: borderColor,
        strokeOpacity: opacity,
        onMouseEnter: handleMouseEnter,
        onMouseMove: handleMouseMove,
        onMouseLeave: handleMouseLeave,
        onClick: onClick
    });
};

SankeyNodesItem.propTypes = {
    node: PropTypes.shape({
        id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
        label: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
        color: PropTypes.string.isRequired
    }),

    x: PropTypes.number.isRequired,
    y: PropTypes.number.isRequired,
    width: PropTypes.number.isRequired,
    height: PropTypes.number.isRequired,

    color: PropTypes.string.isRequired,
    opacity: PropTypes.number.isRequired,
    borderWidth: PropTypes.number.isRequired,
    borderColor: PropTypes.string.isRequired,

    // interactivity
    showTooltip: PropTypes.func.isRequired,
    hideTooltip: PropTypes.func.isRequired,
    setCurrent: PropTypes.func.isRequired,
    onClick: PropTypes.func.isRequired,

    tooltip: PropTypes.element.isRequired,
    theme: PropTypes.object.isRequired
};

var enhance = compose(withPropsOnChange(['node', 'theme', 'tooltip'], function (_ref2) {
    var node = _ref2.node,
        theme = _ref2.theme,
        tooltip = _ref2.tooltip;

    if (tooltip) {
        return {
            tooltip: React.createElement(BasicTooltip, { id: tooltip(node), enableChip: false, theme: theme })
        };
    }
    return {
        tooltip: React.createElement(BasicTooltip, { id: node.label, enableChip: true, color: node.color, theme: theme })
    };
}), withPropsOnChange(['onClick', 'node'], function (_ref3) {
    var _onClick = _ref3.onClick,
        node = _ref3.node;
    return {
        onClick: function onClick(event) {
            return _onClick(node, event);
        }
    };
}), withHandlers({
    handleMouseEnter: function handleMouseEnter(_ref4) {
        var showTooltip = _ref4.showTooltip,
            setCurrent = _ref4.setCurrent,
            node = _ref4.node,
            tooltip = _ref4.tooltip;
        return function (e) {
            setCurrent(node);
            showTooltip(tooltip, e);
        };
    },
    handleMouseMove: function handleMouseMove(_ref5) {
        var showTooltip = _ref5.showTooltip,
            tooltip = _ref5.tooltip;
        return function (e) {
            showTooltip(tooltip, e);
        };
    },
    handleMouseLeave: function handleMouseLeave(_ref6) {
        var hideTooltip = _ref6.hideTooltip,
            setCurrent = _ref6.setCurrent;
        return function () {
            setCurrent(null);
            hideTooltip();
        };
    }
}), pure);

export default enhance(SankeyNodesItem);
