// nivo/es/components/charts/treemap/props.js
var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

/*
 * This file is part of the nivo project.
 *
 * Copyright 2016-present, Raphaël Benitte.
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */
import PropTypes from 'prop-types';
import noop from '../../../lib/noop';
import { treeMapTilePropType, defsPropTypes } from '../../../props';
import TreeMapNode from './TreeMapNode';
import TreeMapHtmlNode from './TreeMapHtmlNode';

/*—————————————————————————————————————————————————————————————————————————————

  Prop types

—————————————————————————————————————————————————————————————————————————————*/

var commonPropTypes = {
    // data
    // `root` managed by `withHierarchy()` HOC
    identity: PropTypes.oneOfType([PropTypes.string, PropTypes.func]).isRequired,

    // dimensions managed by `withDimensions()` HOC

    // styling
    // theme managed by `withTheme()` HOC
    // colors managed by `withColors()` HOC

    leavesOnly: PropTypes.bool.isRequired,
    tile: treeMapTilePropType.isRequired,
    innerPadding: PropTypes.number.isRequired,
    outerPadding: PropTypes.number.isRequired,

    // labels
    enableLabel: PropTypes.bool.isRequired,
    label: PropTypes.oneOfType([PropTypes.string, PropTypes.func]).isRequired,
    labelFormat: PropTypes.string,
    labelSkipSize: PropTypes.number.isRequired,
    orientLabel: PropTypes.bool.isRequired,

    // border
    borderWidth: PropTypes.number.isRequired,
    borderColor: PropTypes.any.isRequired,

    // interactivity
    isInteractive: PropTypes.bool.isRequired,
    onClick: PropTypes.func.isRequired
};

export var TreeMapPropTypes = _extends({}, commonPropTypes, {
    nodeComponent: PropTypes.func.isRequired
}, defsPropTypes);

export var TreeMapPlaceholdersPropTypes = TreeMapPropTypes;

export var TreeMapHtmlPropTypes = _extends({}, commonPropTypes, {
    nodeComponent: PropTypes.func.isRequired
});

export var TreeMapHtmlCanvasPropTypes = _extends({}, commonPropTypes, {
    pixelRatio: PropTypes.number.isRequired

    /*—————————————————————————————————————————————————————————————————————————————
    
      Default props
    
    —————————————————————————————————————————————————————————————————————————————*/

});var commonDefaultProps = {
    // data
    identity: 'id',

    tile: 'squarify',
    leavesOnly: false,

    // labels
    enableLabel: true,
    label: 'id',
    labelSkipSize: 0,
    labelTextColor: 'inherit:darker(1)',
    orientLabel: true,

    innerPadding: 0,
    outerPadding: 0,

    borderWidth: 0,
    borderColor: 'inherit',

    // interactivity
    isInteractive: true,
    onClick: noop
};

export var TreeMapDefaultProps = _extends({}, commonDefaultProps, {
    nodeComponent: TreeMapNode,
    defs: [],
    fill: []
});

export var TreeMapPlaceholdersDefaultProps = TreeMapDefaultProps;

export var TreeMapHtmlDefaultProps = _extends({}, commonDefaultProps, {
    nodeComponent: TreeMapHtmlNode
});

export var TreeMapCanvasDefaultProps = _extends({}, commonDefaultProps, {
    pixelRatio: global.window && global.window.devicePixelRatio ? global.window.devicePixelRatio : 1
});
