// nivo/es/components/charts/treemap/TreeMap.js
var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

/*
 * This file is part of the nivo project.
 *
 * Copyright 2016-present, Raphaël Benitte.
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */
import React from 'react';
import { TransitionMotion, spring } from 'react-motion';
import { colorMotionSpring, getInterpolatedColor } from '../../../lib/colors';
import Container from '../Container';
import enhance from './enhance';
import { nodeWillEnter, nodeWillLeave } from './motion';
import { getNodeHandlers } from './interactivity';
import SvgWrapper from '../SvgWrapper';

var TreeMap = function TreeMap(_ref) {
    var nodes = _ref.nodes,
        nodeComponent = _ref.nodeComponent,
        margin = _ref.margin,
        outerWidth = _ref.outerWidth,
        outerHeight = _ref.outerHeight,
        theme = _ref.theme,
        borderWidth = _ref.borderWidth,
        getBorderColor = _ref.getBorderColor,
        defs = _ref.defs,
        getLabelTextColor = _ref.getLabelTextColor,
        orientLabel = _ref.orientLabel,
        animate = _ref.animate,
        motionStiffness = _ref.motionStiffness,
        motionDamping = _ref.motionDamping,
        isInteractive = _ref.isInteractive,
        onClick = _ref.onClick;

    var springConfig = {
        stiffness: motionStiffness,
        damping: motionDamping
    };

    var getHandlers = function getHandlers(node, showTooltip, hideTooltip) {
        return getNodeHandlers(node, {
            isInteractive: isInteractive,
            onClick: onClick,
            showTooltip: showTooltip,
            hideTooltip: hideTooltip,
            theme: theme
        });
    };

    return React.createElement(
        Container,
        { isInteractive: isInteractive, theme: theme },
        function (_ref2) {
            var showTooltip = _ref2.showTooltip,
                hideTooltip = _ref2.hideTooltip;
            return React.createElement(
                SvgWrapper,
                { width: outerWidth, height: outerHeight, margin: margin, defs: defs },
                !animate && React.createElement(
                    'g',
                    null,
                    nodes.map(function (node) {
                        return React.createElement(nodeComponent, {
                            key: node.path,
                            node: node,
                            style: {
                                fill: node.fill,
                                x: node.x0,
                                y: node.y0,
                                width: node.width,
                                height: node.height,
                                color: node.color,
                                borderWidth: borderWidth,
                                borderColor: getBorderColor(node),
                                labelTextColor: getLabelTextColor(node),
                                orientLabel: orientLabel
                            },
                            handlers: getHandlers(node, showTooltip, hideTooltip)
                        });
                    })
                ),
                animate && React.createElement(
                    TransitionMotion,
                    {
                        willEnter: nodeWillEnter,
                        willLeave: nodeWillLeave(springConfig),
                        styles: nodes.map(function (node) {
                            return {
                                key: node.path,
                                data: node,
                                style: _extends({
                                    x: spring(node.x, springConfig),
                                    y: spring(node.y, springConfig),
                                    width: spring(node.width, springConfig),
                                    height: spring(node.height, springConfig)
                                }, colorMotionSpring(node.color, springConfig))
                            };
                        })
                    },
                    function (interpolatedStyles) {
                        return React.createElement(
                            'g',
                            null,
                            interpolatedStyles.map(function (_ref3) {
                                var style = _ref3.style,
                                    node = _ref3.data;

                                style.color = getInterpolatedColor(style);

                                return React.createElement(nodeComponent, {
                                    key: node.path,
                                    node: node,
                                    style: _extends({}, style, {
                                        fill: node.fill,
                                        borderWidth: borderWidth,
                                        borderColor: getBorderColor(style),
                                        labelTextColor: getLabelTextColor(style),
                                        orientLabel: orientLabel
                                    }),
                                    handlers: getHandlers(node, showTooltip, hideTooltip)
                                });
                            })
                        );
                    }
                )
            );
        }
    );
};

TreeMap.displayName = 'TreeMap';

export default enhance(TreeMap);
