// nivo/es/components/charts/bubble/BubbleNode.js
var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

/*
 * This file is part of the nivo project.
 *
 * Copyright 2016-present, Raphaël Benitte.
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */
import React from 'react';
import PropTypes from 'prop-types';

var BubbleNode = function BubbleNode(_ref) {
    var node = _ref.node,
        style = _ref.style,
        handlers = _ref.handlers;

    if (style.r <= 0) return null;

    return React.createElement(
        'g',
        { transform: 'translate(' + style.x + ',' + style.y + ')' },
        React.createElement('circle', _extends({
            r: style.r
        }, handlers, {
            fill: style.fill ? style.fill : style.color,
            stroke: style.borderColor,
            strokeWidth: style.borderWidth
        })),
        node.label !== false && React.createElement(
            'text',
            {
                textAnchor: 'middle',
                alignmentBaseline: 'central',
                style: {
                    fill: style.labelTextColor,
                    pointerEvents: 'none'
                }
            },
            node.label
        )
    );
};

BubbleNode.propTypes = {
    node: PropTypes.object.isRequired,
    style: PropTypes.shape({
        r: PropTypes.number.isRequired,
        x: PropTypes.number.isRequired,
        y: PropTypes.number.isRequired,
        color: PropTypes.string.isRequired,
        fill: PropTypes.string,
        borderWidth: PropTypes.number.isRequired,
        borderColor: PropTypes.string.isRequired,
        labelTextColor: PropTypes.string.isRequired
    }).isRequired,
    handlers: PropTypes.object.isRequired
};

export default BubbleNode;
