// nivo/lib/components/charts/calendar/CalendarMonthPath.js
'use strict';

exports.__esModule = true;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _pure = require('recompose/pure');

var _pure2 = _interopRequireDefault(_pure);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var CalendarMonthPath = function CalendarMonthPath(_ref) {
    var path = _ref.path,
        borderWidth = _ref.borderWidth,
        borderColor = _ref.borderColor;
    return _react2.default.createElement('path', {
        d: path,
        style: {
            fill: 'none',
            strokeWidth: borderWidth,
            stroke: borderColor
        }
    });
}; /*
    * This file is part of the nivo project.
    *
    * Copyright 2016-present, Raphaël Benitte.
    *
    * For the full copyright and license information, please view the LICENSE
    * file that was distributed with this source code.
    */


CalendarMonthPath.propTypes = {
    path: _propTypes2.default.string.isRequired,
    borderWidth: _propTypes2.default.number.isRequired,
    borderColor: _propTypes2.default.string.isRequired
};

exports.default = (0, _pure2.default)(CalendarMonthPath);
