// nivo/lib/components/charts/calendar/CalendarDay.js
'use strict';

exports.__esModule = true;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _compose = require('recompose/compose');

var _compose2 = _interopRequireDefault(_compose);

var _withPropsOnChange = require('recompose/withPropsOnChange');

var _withPropsOnChange2 = _interopRequireDefault(_withPropsOnChange);

var _pure = require('recompose/pure');

var _pure2 = _interopRequireDefault(_pure);

var _noop = require('../../../lib/noop');

var _noop2 = _interopRequireDefault(_noop);

var _BasicTooltip = require('../../tooltip/BasicTooltip');

var _BasicTooltip2 = _interopRequireDefault(_BasicTooltip);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var CalendarDay = function CalendarDay(_ref) {
    var x = _ref.x,
        y = _ref.y,
        size = _ref.size,
        color = _ref.color,
        borderWidth = _ref.borderWidth,
        borderColor = _ref.borderColor,
        onClick = _ref.onClick,
        showTooltip = _ref.showTooltip,
        hideTooltip = _ref.hideTooltip;
    return _react2.default.createElement('rect', {
        x: x,
        y: y,
        width: size,
        height: size,
        style: {
            fill: color,
            strokeWidth: borderWidth,
            stroke: borderColor
        },
        onClick: onClick,
        onMouseEnter: showTooltip,
        onMouseMove: showTooltip,
        onMouseLeave: hideTooltip
    });
}; /*
    * This file is part of the nivo project.
    *
    * Copyright 2016-present, Raphaël Benitte.
    *
    * For the full copyright and license information, please view the LICENSE
    * file that was distributed with this source code.
    */


CalendarDay.propTypes = {
    onClick: _propTypes2.default.func.isRequired,
    data: _propTypes2.default.object.isRequired,
    x: _propTypes2.default.number.isRequired,
    y: _propTypes2.default.number.isRequired,
    size: _propTypes2.default.number.isRequired,
    color: _propTypes2.default.string.isRequired,
    borderWidth: _propTypes2.default.number.isRequired,
    borderColor: _propTypes2.default.string.isRequired,

    tooltipFormat: _propTypes2.default.oneOfType([_propTypes2.default.func, _propTypes2.default.string]),
    showTooltip: _propTypes2.default.func.isRequired,
    hideTooltip: _propTypes2.default.func.isRequired,

    theme: _propTypes2.default.shape({
        tooltip: _propTypes2.default.shape({}).isRequired
    }).isRequired
};

var enhance = (0, _compose2.default)((0, _withPropsOnChange2.default)(['data', 'onClick'], function (_ref2) {
    var data = _ref2.data,
        _onClick = _ref2.onClick;
    return {
        onClick: function onClick(event) {
            return _onClick(data, event);
        }
    };
}), (0, _withPropsOnChange2.default)(['data', 'color', 'showTooltip', 'theme', 'tooltipFormat'], function (_ref3) {
    var data = _ref3.data,
        color = _ref3.color,
        _showTooltip = _ref3.showTooltip,
        theme = _ref3.theme,
        tooltipFormat = _ref3.tooltipFormat;

    if (data.value === undefined) return { showTooltip: _noop2.default };

    return {
        showTooltip: function showTooltip(event) {
            return _showTooltip(_react2.default.createElement(_BasicTooltip2.default, {
                id: '' + data.day,
                value: data.value,
                enableChip: true,
                color: color,
                theme: theme,
                format: tooltipFormat
            }), event);
        }
    };
}), _pure2.default);

exports.default = enhance(CalendarDay);
