// nivo/lib/components/charts/calendar/props.js
'use strict';

exports.__esModule = true;
exports.CalendarDefaultProps = exports.CalendarPropTypes = undefined;

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _noop = require('../../../lib/noop');

var _noop2 = _interopRequireDefault(_noop);

var _directions = require('../../../constants/directions');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Calendar components propTypes.
 *
 * @type {object}
 */
var CalendarPropTypes = exports.CalendarPropTypes = {
    from: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.instanceOf(Date)]).isRequired,
    to: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.instanceOf(Date)]).isRequired,
    data: _propTypes2.default.arrayOf(_propTypes2.default.shape({
        day: _propTypes2.default.string.isRequired,
        value: _propTypes2.default.number.isRequired
    })).isRequired,

    domain: _propTypes2.default.oneOfType([_propTypes2.default.oneOf(['auto']), _propTypes2.default.arrayOf(_propTypes2.default.number)]).isRequired,
    colors: _propTypes2.default.arrayOf(_propTypes2.default.string).isRequired,
    colorScale: _propTypes2.default.func.isRequired,

    onDayClick: _propTypes2.default.func.isRequired,
    direction: _propTypes2.default.oneOf([_directions.DIRECTION_HORIZONTAL, _directions.DIRECTION_VERTICAL]),
    emptyColor: _propTypes2.default.string.isRequired,

    // years
    yearSpacing: _propTypes2.default.number.isRequired,
    yearLegendOffset: _propTypes2.default.number.isRequired,

    // months
    monthBorderWidth: _propTypes2.default.number.isRequired,
    monthBorderColor: _propTypes2.default.string.isRequired,
    monthLegendOffset: _propTypes2.default.number.isRequired,

    // days
    daySpacing: _propTypes2.default.number.isRequired,
    dayBorderWidth: _propTypes2.default.number.isRequired,
    dayBorderColor: _propTypes2.default.string.isRequired,

    // interactivity
    isInteractive: _propTypes2.default.bool,
    onClick: _propTypes2.default.func.isRequired,
    tooltipFormat: _propTypes2.default.oneOfType([_propTypes2.default.func, _propTypes2.default.string])

    /**
     * Calendar components defaultProps.
     *
     * @type {object}
     */
}; /*
    * This file is part of the nivo project.
    *
    * Copyright 2016-present, Raphaël Benitte.
    *
    * For the full copyright and license information, please view the LICENSE
    * file that was distributed with this source code.
    */
var CalendarDefaultProps = exports.CalendarDefaultProps = {
    domain: 'auto',
    colors: ['#61cdbb', '#97e3d5', '#e8c1a0', '#f47560'],

    direction: _directions.DIRECTION_HORIZONTAL,
    onDayClick: function onDayClick() {},
    emptyColor: '#fff',

    // years
    yearSpacing: 30,
    yearLegendOffset: 10,

    // months
    monthBorderWidth: 2,
    monthBorderColor: '#000',
    monthLegendOffset: 6,

    // days
    daySpacing: 0,
    dayBorderWidth: 1,
    dayBorderColor: '#000',

    // interactivity
    isInteractive: true,
    onClick: _noop2.default
};
