// nivo/lib/components/charts/bubble/BubbleHtmlNode.js
'use strict';

exports.__esModule = true;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; /*
                                                                                                                                                                                                                                                                   * This file is part of the nivo project.
                                                                                                                                                                                                                                                                   *
                                                                                                                                                                                                                                                                   * Copyright 2016-present, Raphaël Benitte.
                                                                                                                                                                                                                                                                   *
                                                                                                                                                                                                                                                                   * For the full copyright and license information, please view the LICENSE
                                                                                                                                                                                                                                                                   * file that was distributed with this source code.
                                                                                                                                                                                                                                                                   */


var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var BubbleHtmlNode = function BubbleHtmlNode(_ref) {
    var node = _ref.node,
        style = _ref.style,
        handlers = _ref.handlers;

    if (style.r <= 0) return null;

    return _react2.default.createElement(
        'div',
        _extends({
            style: {
                position: 'absolute',
                display: 'flex',
                alignItems: 'center',
                justifyContent: 'center',
                background: style.color,
                color: style.labelTextColor,
                borderWidth: style.borderWidth,
                borderColor: style.borderColor,
                top: style.y - style.r,
                left: style.x - style.r,
                width: style.r * 2,
                height: style.r * 2,
                borderStyle: 'solid',
                borderRadius: style.r
            }
        }, handlers),
        node.label !== false && node.label
    );
};

BubbleHtmlNode.propTypes = {
    node: _propTypes2.default.object.isRequired,
    style: _propTypes2.default.shape({
        r: _propTypes2.default.number.isRequired,
        x: _propTypes2.default.number.isRequired,
        y: _propTypes2.default.number.isRequired,
        color: _propTypes2.default.string.isRequired,
        fill: _propTypes2.default.string,
        borderWidth: _propTypes2.default.number.isRequired,
        borderColor: _propTypes2.default.string.isRequired,
        labelTextColor: _propTypes2.default.string.isRequired
    }).isRequired,
    handlers: _propTypes2.default.object.isRequired
};

exports.default = BubbleHtmlNode;
