// nivo/lib/components/charts/bubble/enhance.js
'use strict';

exports.__esModule = true;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; /*
                                                                                                                                                                                                                                                                   * This file is part of the nivo project.
                                                                                                                                                                                                                                                                   *
                                                                                                                                                                                                                                                                   * Copyright 2016-present, Raphaël Benitte.
                                                                                                                                                                                                                                                                   *
                                                                                                                                                                                                                                                                   * For the full copyright and license information, please view the LICENSE
                                                                                                                                                                                                                                                                   * file that was distributed with this source code.
                                                                                                                                                                                                                                                                   */


var _d3Hierarchy = require('d3-hierarchy');

var _compose = require('recompose/compose');

var _compose2 = _interopRequireDefault(_compose);

var _defaultProps = require('recompose/defaultProps');

var _defaultProps2 = _interopRequireDefault(_defaultProps);

var _withPropsOnChange = require('recompose/withPropsOnChange');

var _withPropsOnChange2 = _interopRequireDefault(_withPropsOnChange);

var _withStateHandlers = require('recompose/withStateHandlers');

var _withStateHandlers2 = _interopRequireDefault(_withStateHandlers);

var _pure = require('recompose/pure');

var _pure2 = _interopRequireDefault(_pure);

var _hocs = require('../../../hocs');

var _propertiesConverters = require('../../../lib/propertiesConverters');

var _colors = require('../../../lib/colors');

var _defs = require('../../../lib/defs');

var _bubble = require('../../../lib/charts/bubble');

var _props = require('./props');

var props = _interopRequireWildcard(_props);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var commonEnhancers = [(0, _hocs.withHierarchy)(), (0, _hocs.withDimensions)(), (0, _hocs.withTheme)(), (0, _hocs.withColors)({ defaultColorBy: 'depth' }), (0, _withPropsOnChange2.default)(['width', 'height', 'padding'], function (_ref) {
    var width = _ref.width,
        height = _ref.height,
        padding = _ref.padding;
    return {
        pack: (0, _d3Hierarchy.pack)().size([width, height]).padding(padding)
    };
}), (0, _withPropsOnChange2.default)(['identity'], function (_ref2) {
    var identity = _ref2.identity;
    return {
        getIdentity: (0, _propertiesConverters.getAccessorFor)(identity)
    };
}),

// border
(0, _withPropsOnChange2.default)(['borderColor'], function (_ref3) {
    var borderColor = _ref3.borderColor;
    return {
        getBorderColor: (0, _colors.getInheritedColorGenerator)(borderColor)
    };
}),

// labels
(0, _withPropsOnChange2.default)(['label', 'labelFormat'], function (_ref4) {
    var label = _ref4.label,
        labelFormat = _ref4.labelFormat;
    return {
        getLabel: (0, _propertiesConverters.getLabelGenerator)(label, labelFormat)
    };
}), (0, _withPropsOnChange2.default)(['labelTextColor'], function (_ref5) {
    var labelTextColor = _ref5.labelTextColor;
    return {
        getLabelTextColor: (0, _colors.getInheritedColorGenerator)(labelTextColor)
    };
}),

// zoom
(0, _withStateHandlers2.default)(function (_ref6) {
    var _ref6$currentNodePath = _ref6.currentNodePath,
        currentNodePath = _ref6$currentNodePath === undefined ? null : _ref6$currentNodePath;
    return {
        currentNodePath: currentNodePath
    };
}, {
    zoomToNode: function zoomToNode(_ref7) {
        var currentNodePath = _ref7.currentNodePath;
        return function (path) {
            if (path === currentNodePath) return { currentNodePath: null };
            return { currentNodePath: path };
        };
    }
}), (0, _withPropsOnChange2.default)(['root', 'pack', 'leavesOnly', 'getIdentity', 'getColor'], function (_ref8) {
    var root = _ref8.root,
        pack = _ref8.pack,
        leavesOnly = _ref8.leavesOnly,
        getIdentity = _ref8.getIdentity,
        getColor = _ref8.getColor;

    var nodes = (0, _bubble.computeNodes)({ root: root, pack: pack, leavesOnly: leavesOnly, getIdentity: getIdentity, getColor: getColor });

    return { nodes: nodes };
}), (0, _withPropsOnChange2.default)(['enableLabel', 'nodes', 'getLabel', 'labelSkipRadius'], function (_ref9) {
    var enableLabel = _ref9.enableLabel,
        nodes = _ref9.nodes,
        getLabel = _ref9.getLabel,
        labelSkipRadius = _ref9.labelSkipRadius;

    if (!enableLabel) return;
    var nodesWithLabel = nodes.map(function (node) {
        if (node.height !== 0 || labelSkipRadius > 0 && node.r < labelSkipRadius) return node;
        return _extends({}, node, { label: getLabel(node) });
    });

    return { nodes: nodesWithLabel };
}), (0, _withPropsOnChange2.default)(['nodes', 'isZoomable', 'currentNodePath'], function (_ref10) {
    var nodes = _ref10.nodes,
        isZoomable = _ref10.isZoomable,
        currentNodePath = _ref10.currentNodePath,
        width = _ref10.width,
        height = _ref10.height;

    if (currentNodePath && isZoomable) {
        return {
            nodes: (0, _bubble.computeZoom)(nodes, currentNodePath, width, height)
        };
    }
})];

var svgEnhancers = [(0, _withPropsOnChange2.default)(['nodes', 'defs', 'fill'], function (_ref11) {
    var nodes = _ref11.nodes,
        defs = _ref11.defs,
        fill = _ref11.fill;

    return {
        defs: (0, _defs.bindDefs)(defs, nodes, fill, { targetKey: 'fill' })
    };
})];

exports.default = function (Component) {
    var implPropTypes = props[Component.displayName + 'PropTypes'];
    var implDefaultProps = props[Component.displayName + 'DefaultProps'];

    Component.propTypes = implPropTypes;

    switch (Component.displayName) {
        case 'Bubble':
            return _compose2.default.apply(undefined, [(0, _defaultProps2.default)(implDefaultProps)].concat(commonEnhancers, svgEnhancers, [(0, _hocs.withMotion)(), _pure2.default]))(Component);

        case 'BubbleHtml':
            return _compose2.default.apply(undefined, [(0, _defaultProps2.default)(implDefaultProps)].concat(commonEnhancers, [(0, _hocs.withMotion)(), _pure2.default]))(Component);

        case 'BubbleCanvas':
            return _compose2.default.apply(undefined, [(0, _defaultProps2.default)(implDefaultProps)].concat(commonEnhancers, [_pure2.default]))(Component);
    }

    return Component;
};
