// nivo/lib/components/charts/bubble/motion.js
'use strict';

exports.__esModule = true;
exports.nodeWillLeave = exports.nodeWillEnter = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; /*
                                                                                                                                                                                                                                                                   * This file is part of the nivo project.
                                                                                                                                                                                                                                                                   *
                                                                                                                                                                                                                                                                   * Copyright 2016-present, Raphaël Benitte.
                                                                                                                                                                                                                                                                   *
                                                                                                                                                                                                                                                                   * For the full copyright and license information, please view the LICENSE
                                                                                                                                                                                                                                                                   * file that was distributed with this source code.
                                                                                                                                                                                                                                                                   */


var _reactMotion = require('react-motion');

var _colors = require('../../../lib/colors');

var nodeWillEnter = exports.nodeWillEnter = function nodeWillEnter(_ref) {
    var data = _ref.data;
    return _extends({
        scale: 0,
        r: 0,
        x: data.x,
        y: data.y
    }, (0, _colors.colorMotionSpring)(data.color));
};

var nodeWillLeave = exports.nodeWillLeave = function nodeWillLeave(springConfig) {
    return function (_ref2) {
        var data = _ref2.data;
        return _extends({
            scale: (0, _reactMotion.spring)(0, springConfig),
            r: (0, _reactMotion.spring)(0, springConfig),
            x: (0, _reactMotion.spring)(data.x, springConfig),
            y: (0, _reactMotion.spring)(data.y, springConfig)
        }, (0, _colors.colorMotionSpring)(data.color, springConfig));
    };
};
