// nivo/lib/components/charts/bubble/props.js
'use strict';

exports.__esModule = true;
exports.BubbleCanvasDefaultProps = exports.BubbleHtmlDefaultProps = exports.BubbleDefaultProps = exports.BubbleCanvasPropTypes = exports.BubbleHtmlPropTypes = exports.BubblePropTypes = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; /*
                                                                                                                                                                                                                                                                   * This file is part of the nivo project.
                                                                                                                                                                                                                                                                   *
                                                                                                                                                                                                                                                                   * Copyright 2016-present, Raphaël Benitte.
                                                                                                                                                                                                                                                                   *
                                                                                                                                                                                                                                                                   * For the full copyright and license information, please view the LICENSE
                                                                                                                                                                                                                                                                   * file that was distributed with this source code.
                                                                                                                                                                                                                                                                   */


var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _noop = require('../../../lib/noop');

var _noop2 = _interopRequireDefault(_noop);

var _props = require('../../../props');

var _BubbleNode = require('./BubbleNode');

var _BubbleNode2 = _interopRequireDefault(_BubbleNode);

var _BubbleHtmlNode = require('./BubbleHtmlNode');

var _BubbleHtmlNode2 = _interopRequireDefault(_BubbleHtmlNode);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/*—————————————————————————————————————————————————————————————————————————————

  Prop types

—————————————————————————————————————————————————————————————————————————————*/

var commonPropTypes = {
    // data
    // `root` managed by `withHierarchy()` HOC
    identity: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.func]).isRequired,

    // dimensions managed by `withDimensions()` HOC

    // styling
    // theme managed by `withTheme()` HOC
    // colors managed by `withColors()` HOC

    leavesOnly: _propTypes2.default.bool.isRequired,
    padding: _propTypes2.default.number.isRequired,

    // border
    borderWidth: _propTypes2.default.number.isRequired,
    borderColor: _propTypes2.default.any.isRequired,

    // labels
    enableLabel: _propTypes2.default.bool.isRequired,
    label: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.func]).isRequired,
    labelFormat: _propTypes2.default.string,
    labelTextColor: _propTypes2.default.any.isRequired,
    labelSkipRadius: _propTypes2.default.number.isRequired,

    // interactivity
    isInteractive: _propTypes2.default.bool.isRequired,
    onClick: _propTypes2.default.func.isRequired,
    isZoomable: _propTypes2.default.bool.isRequired,
    tooltipFormat: _propTypes2.default.oneOfType([_propTypes2.default.func, _propTypes2.default.string])
};

var BubblePropTypes = exports.BubblePropTypes = _extends({}, commonPropTypes, {
    nodeComponent: _propTypes2.default.func.isRequired
}, _props.defsPropTypes);

var BubbleHtmlPropTypes = exports.BubbleHtmlPropTypes = _extends({}, commonPropTypes, {
    nodeComponent: _propTypes2.default.func.isRequired
});

var BubbleCanvasPropTypes = exports.BubbleCanvasPropTypes = _extends({}, commonPropTypes, {
    pixelRatio: _propTypes2.default.number.isRequired

    /*—————————————————————————————————————————————————————————————————————————————
    
      Default props
    
    —————————————————————————————————————————————————————————————————————————————*/

});var commonDefaultProps = {
    identity: 'id',

    leavesOnly: false,
    padding: 1,

    // border
    borderWidth: 0,
    borderColor: 'inherit',

    // labels
    enableLabel: true,
    label: 'id',
    labelTextColor: 'inherit:darker(1)',
    labelSkipRadius: 8,

    // interactivity
    isInteractive: true,
    onClick: _noop2.default,
    isZoomable: true
};

var BubbleDefaultProps = exports.BubbleDefaultProps = _extends({}, commonDefaultProps, {
    nodeComponent: _BubbleNode2.default,
    defs: [],
    fill: []
});

var BubbleHtmlDefaultProps = exports.BubbleHtmlDefaultProps = _extends({}, commonDefaultProps, {
    nodeComponent: _BubbleHtmlNode2.default
});

var BubbleCanvasDefaultProps = exports.BubbleCanvasDefaultProps = _extends({}, commonDefaultProps, {
    pixelRatio: global.window && global.window.devicePixelRatio ? global.window.devicePixelRatio : 1
});
