// nivo/lib/components/charts/voronoi/props.js
'use strict';

exports.__esModule = true;
exports.VoronoiDefaultProps = exports.VoronoiPropTypes = undefined;

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var VoronoiPropTypes = exports.VoronoiPropTypes = {
    data: _propTypes2.default.arrayOf(_propTypes2.default.shape({
        id: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.number]).isRequired,
        x: _propTypes2.default.number.isRequired,
        y: _propTypes2.default.number.isRequired
    })).isRequired,
    enablePolygons: _propTypes2.default.bool.isRequired,
    enableSites: _propTypes2.default.bool.isRequired,
    enableLinks: _propTypes2.default.bool.isRequired,
    x: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.number]).isRequired,
    y: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.number]).isRequired,
    borderWidth: _propTypes2.default.number.isRequired,
    borderColor: _propTypes2.default.string.isRequired,
    linkWidth: _propTypes2.default.number.isRequired,
    linkColor: _propTypes2.default.string.isRequired,
    siteSize: _propTypes2.default.number.isRequired,
    siteColor: _propTypes2.default.string.isRequired
}; /*
    * This file is part of the nivo project.
    *
    * Copyright 2016-present, Raphaël Benitte.
    *
    * For the full copyright and license information, please view the LICENSE
    * file that was distributed with this source code.
    */
var VoronoiDefaultProps = exports.VoronoiDefaultProps = {
    enablePolygons: true,
    enableSites: false,
    enableLinks: false,
    x: 0,
    y: 1,
    borderWidth: 2,
    borderColor: '#000',
    linkWidth: 1,
    linkColor: '#bbb',
    siteSize: 4,
    siteColor: '#666'
};
