// nivo/lib/components/charts/treemap/interactivity.js
'use strict';

exports.__esModule = true;
exports.getNodeHandlers = undefined;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _TreeMapNodeTooltip = require('./TreeMapNodeTooltip');

var _TreeMapNodeTooltip2 = _interopRequireDefault(_TreeMapNodeTooltip);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/*
 * This file is part of the nivo project.
 *
 * Copyright 2016-present, Raphaël Benitte.
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */
var getNodeHandlers = exports.getNodeHandlers = function getNodeHandlers(node, _ref) {
    var isInteractive = _ref.isInteractive,
        _onClick = _ref.onClick,
        showTooltip = _ref.showTooltip,
        hideTooltip = _ref.hideTooltip,
        theme = _ref.theme;

    if (!isInteractive) return {};

    var handleTooltip = function handleTooltip(e) {
        showTooltip(_react2.default.createElement(_TreeMapNodeTooltip2.default, { node: node, theme: theme }), e);
    };

    return {
        onMouseEnter: handleTooltip,
        onMouseMove: handleTooltip,
        onMouseLeave: hideTooltip,
        onClick: function onClick(event) {
            return _onClick(node, event);
        }
    };
};
