// nivo/lib/components/charts/treemap/motion.js
'use strict';

exports.__esModule = true;
exports.nodeWillLeave = exports.nodeWillEnter = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; /*
                                                                                                                                                                                                                                                                   * This file is part of the nivo project.
                                                                                                                                                                                                                                                                   *
                                                                                                                                                                                                                                                                   * Copyright 2016-present, Raphaël Benitte.
                                                                                                                                                                                                                                                                   *
                                                                                                                                                                                                                                                                   * For the full copyright and license information, please view the LICENSE
                                                                                                                                                                                                                                                                   * file that was distributed with this source code.
                                                                                                                                                                                                                                                                   */


var _reactMotion = require('react-motion');

var _colors = require('../../../lib/colors');

var nodeWillEnter = exports.nodeWillEnter = function nodeWillEnter(_ref) {
    var node = _ref.data;
    return _extends({
        x: node.x,
        y: node.y,
        width: node.width,
        height: node.height
    }, (0, _colors.colorMotionSpring)(node.color));
};

var nodeWillLeave = exports.nodeWillLeave = function nodeWillLeave(springConfig) {
    return function (_ref2) {
        var node = _ref2.data;
        return _extends({
            x: (0, _reactMotion.spring)(node.x + node.width / 2, springConfig),
            y: (0, _reactMotion.spring)(node.y + node.height / 2, springConfig),
            width: (0, _reactMotion.spring)(0, springConfig),
            height: (0, _reactMotion.spring)(0, springConfig)
        }, (0, _colors.colorMotionSpring)(node.color, springConfig));
    };
};
